File: mobile/lib/features/user/data/repositories/user_repository.dart
import 'package:mobile/features/user/data/api/user_api.dart';
import 'package:mobile/features/auth/auth.dart';

class UserRepository {
  UserRepository({required UserApi userApi}) : _userApi = userApi;

  final UserApi _userApi;

  Future<User> updateProfile({
    required String firstName,
    required String lastName,
    required String phone,
  }) {
    return _userApi.updateProfile(
      firstName: firstName,
      lastName: lastName,
      phone: phone,
    );
  }

  Future<void> updatePassword({
    required String currentPassword,
    required String newPassword,
    required String confirmNewPassword,
  }) {
    return _userApi.updatePassword(
      currentPassword: currentPassword,
      newPassword: newPassword,
      confirmNewPassword: confirmNewPassword,
    );
  }

  Future<void> requestEmailChangeOtp() {
    return _userApi.requestEmailChangeOtp();
  }

  Future<String> verifyEmailChangeOtp(String otp) {
    return _userApi.verifyEmailChangeOtp(otp);
  }

  Future<User> updateEmail({
    required String newEmail,
    required String changeEmailToken,
  }) {
    return _userApi.updateEmail(
      newEmail: newEmail,
      changeEmailToken: changeEmailToken,
    );
  }

  Future<User> joinShopByCode(String code) {
    return _userApi.joinShopByCode(code);
  }

  Future<User> unlinkFromShop() {
    return _userApi.unlinkFromShop();
  }
}

File: web/src/app/features/user/data/repositories/user.repository.ts
import { Injectable, inject } from '@angular/core';

import { SessionStore } from '../../../../core/store/session.store';
import { UserApiService } from '../api/user-api.service';
import type { UserProfile } from '../../domain/entities/user-profile';

@Injectable({ providedIn: 'root' })
export class UserRepository {
  private readonly userApi = inject(UserApiService);
  private readonly sessionStore = inject(SessionStore);

  getCurrentProfile(): UserProfile | null {
    const user = this.sessionStore.user();
    if (!user) {
      return null;
    }

    return {
      id: user.id,
      first_name: user.first_name,
      last_name: user.last_name,
      email: user.email,
      phone: user.phone,
      role: user.role,
    };
  }

  updateProfile(data: unknown) {
    return this.userApi.updateProfile(data);
  }

  updatePassword(data: unknown) {
    return this.userApi.updatePassword(data);
  }

  requestEmailChangeOtp() {
    return this.userApi.requestEmailChangeOtp();
  }

  verifyEmailChangeOtp(data: unknown) {
    return this.userApi.verifyEmailChangeOtp(data);
  }

  updateEmail(data: unknown) {
    return this.userApi.updateEmail(data);
  }
}
